File: sdks/dart/lib/src/session.dart
/// Sibna Protocol v3.0.1 — Session Management
///
/// High-level session that combines X3DH handshake with Double Ratchet
/// encryption. Matches the Rust core's DoubleRatchetSession API.
import 'dart:typed_data';

import 'crypto.dart';
import 'ratchet.dart';
import 'x3dh.dart';

// ── Session Config ──────────────────────────────────────────────────────────

/// Session configuration.
class SessionConfig {
  final int maxSkippedMessages;
  final int maxChainMessages;

  SessionConfig({
    int? maxSkippedMessages,
    int? maxChainMessages,
  })  : maxSkippedMessages = maxSkippedMessages ?? 2000,
        maxChainMessages = maxChainMessages ?? 4000;
}

// ── Session ─────────────────────────────────────────────────────────────────

/// Encrypted session combining X3DH + Double Ratchet.
class Session {
  DoubleRatchet? _ratchet;
  final String _sessionId;
  DateTime? _establishedAt;
  Uint8List? _peerId;

  Session({SessionConfig? config}) : _sessionId = _generateSessionId();

  /// Unique session identifier.
  String get sessionId => _sessionId;

  /// Whether the session is established (ratchet initialized).
  bool get isEstablished => _ratchet != null;

  /// When the session was established.
  DateTime? get establishedAt => _establishedAt;

  /// The peer's identity key (Ed25519 public key bytes).
  Uint8List? get peerId => _peerId;

  /// Number of messages sent.
  int get messagesSent => _ratchet?.messagesSent ?? 0;

  /// Number of messages received.
  int get messagesReceived => _ratchet?.messagesReceived ?? 0;

  // ── Initiator (who creates the session first) ───────────────────────

  /// Perform X3DH as initiator and initialize the Double Ratchet.
  ///
  /// Returns the ephemeral public key (sent to responder).
  Future<Uint8List> initiateAsInitiator({
    required X25519KeyPair identity,
    required X25519KeyPair ephemeral,
    required PreKeyBundle peerBundle,
    Uint8List? ourDeviceId,
    Uint8List? peerDeviceId,
  }) async {
    final result = await x3DHInitiator(
      identityKeypair: identity,
      ephemeralKeypair: ephemeral,
      peerBundle: peerBundle,
      ourDeviceId: ourDeviceId,
      peerDeviceId: peerDeviceId,
    );

    final ephemeralPrivateBytes = await ephemeral.privateKey;

    _ratchet = await DoubleRatchet.fromSharedSecret(
      sharedSecret: result.sharedSecret,
      localDhPrivate: ephemeralPrivateBytes,
      remoteDhPublic: peerBundle.signedPrekey,
      roleIsInitiator: true,
    );
    _establishedAt = DateTime.now();
    _peerId = peerBundle.identityKey;

    return ephemeral.publicKey;
  }

  // ── Responder (who receives the first message) ──────────────────────

  /// Perform X3DH as responder and initialize the Double Ratchet.
  Future<void> initiateAsResponder({
    required X25519KeyPair identity,
    required X25519KeyPair signedPrekey,
    X25519KeyPair? onetimePrekey,
    required Uint8List peerIdentity,
    required Uint8List peerEphemeral,
    Uint8List? ourDeviceId,
    Uint8List? peerDeviceId,
  }) async {
    final result = await x3DHResponder(
      identityKeypair: identity,
      signedPrekeypair: signedPrekey,
      onetimePrekeypair: onetimePrekey,
      peerIdentity: peerIdentity,
      peerEphemeral: peerEphemeral,
      ourDeviceId: ourDeviceId,
      peerDeviceId: peerDeviceId,
    );

    final signedPrekeyPrivateBytes = await signedPrekey.privateKey;

    _ratchet = await DoubleRatchet.fromSharedSecret(
      sharedSecret: result.sharedSecret,
      localDhPrivate: signedPrekeyPrivateBytes,
      remoteDhPublic: peerEphemeral,
      roleIsInitiator: false,
    );
    _establishedAt = DateTime.now();
    _peerId = peerIdentity;
  }

  // ── Restore from known shared secret (testing) ──────────────────────

  /// Restore a session from a known shared secret (for testing).
  static Future<Session> fromSharedSecret({
    required Uint8List sharedSecret,
    required Uint8List localDhPrivate,
    required Uint8List remoteDhPublic,
    required bool roleIsInitiator,
    SessionConfig? config,
  }) async {
    final s = Session(config: config);
    s._ratchet = await DoubleRatchet.fromSharedSecret(
      sharedSecret: sharedSecret,
      localDhPrivate: localDhPrivate,
      remoteDhPublic: remoteDhPublic,
      roleIsInitiator: roleIsInitiator,
    );
    s._establishedAt = DateTime.now();
    return s;
  }

  // ── Encrypt / Decrypt ───────────────────────────────────────────────

  /// Encrypt a message.
  Future<Uint8List> encrypt(
    Uint8List plaintext, {
    Uint8List? associatedData,
  }) async {
    if (_ratchet == null) throw StateError('Session not established');
    return _ratchet!.ratchetEncrypt(plaintext, associatedData: associatedData);
  }

  /// Decrypt a message.
  Future<Uint8List> decrypt(
    Uint8List ciphertext, {
    Uint8List? associatedData,
  }) async {
    if (_ratchet == null) throw StateError('Session not established');
    return _ratchet!.ratchetDecrypt(ciphertext, associatedData: associatedData);
  }

  // ── Serialization ───────────────────────────────────────────────────

  /// Export session state for persistence.
  Map<String, dynamic>? exportState() {
    if (_ratchet == null) return null;
    final r = _ratchet!;
    return {
      'sessionId': _sessionId,
      'rootKey': bytesToHex(r.rootKey),
      'dhLocalPrivate':
          r.dhLocalPrivate != null ? bytesToHex(r.dhLocalPrivate!) : null,
      'dhLocalPublic':
          r.dhLocalPublic != null ? bytesToHex(r.dhLocalPublic!) : null,
      'dhRemotePublic':
          r.dhRemotePublic != null ? bytesToHex(r.dhRemotePublic!) : null,
      'sendingChainKey':
          r.sendingChain != null ? bytesToHex(r.sendingChain!.key) : null,
      'sendingChainIndex': r.sendingChain?.index ?? 0,
      'receivingChainKey':
          r.receivingChain != null ? bytesToHex(r.receivingChain!.key) : null,
      'receivingChainIndex': r.receivingChain?.index ?? 0,
      'messagesSent': r.messagesSent,
      'messagesReceived': r.messagesReceived,
      'previousCounter': r.previousCounter,
    };
  }

  @override
  String toString() {
    final peerHex =
        _peerId != null ? bytesToHex(_peerId!).substring(0, 16) : 'None';
    return 'Session(id: ${_sessionId.substring(0, 8)}..., '
        'peer: $peerHex..., '
        'sent: $messagesSent, recv: $messagesReceived, '
        'established: $isEstablished)';
  }

  // ── Helpers ─────────────────────────────────────────────────────────

  static String _generateSessionId() {
    final bytes = randomBytes(16);
    return bytesToHex(bytes);
  }
}

File: sdks/javascript/src/session.ts
/**
 * Sibna Protocol v3.0.1 — Session (JavaScript)
 */

import { X25519KeyPair } from './crypto';
import { DoubleRatchet } from './ratchet';
import { PreKeyBundle, x3dhInitiator, x3dhResponder } from './x3dh';

export interface SessionConfig {
  maxSkippedMessages?: number;
  maxChainMessages?: number;
}

export class Session {
  private _ratchet: DoubleRatchet | null = null;
  private _sessionId: string;
  private _establishedAt: number | null = null;
  private _peerId: Uint8Array | null = null;

  constructor() {
    this._sessionId = crypto.randomUUID();
  }

  get sessionId(): string { return this._sessionId; }
  get isEstablished(): boolean { return this._ratchet !== null; }
  get establishedAt(): number | null { return this._establishedAt; }
  get peerId(): Uint8Array | null { return this._peerId ? new Uint8Array(this._peerId) : null; }
  get messagesSent(): number { return this._ratchet?.messagesSent ?? 0; }
  get messagesReceived(): number { return this._ratchet?.messagesReceived ?? 0; }

  initiateAsInitiator(
    identity: X25519KeyPair,
    ephemeral: X25519KeyPair,
    peerBundle: PreKeyBundle,
    ourDeviceId: Uint8Array = new Uint8Array(16),
    peerDeviceId: Uint8Array = new Uint8Array(16),
  ): Uint8Array {
    const result = x3dhInitiator(identity, ephemeral, peerBundle, ourDeviceId, peerDeviceId);
    this._ratchet = DoubleRatchet.fromSharedSecret(
      result.sharedSecret, ephemeral.privateKey, peerBundle.signedPrekey, true,
    );
    this._establishedAt = Date.now();
    this._peerId = new Uint8Array(peerBundle.x25519IdentityKey);
    return ephemeral.publicKey;
  }

  initiateAsResponder(
    identity: X25519KeyPair,
    signedPrekey: X25519KeyPair,
    onetimePrekey: X25519KeyPair | null,
    peerIdentity: Uint8Array,
    peerEphemeral: Uint8Array,
    ourDeviceId: Uint8Array = new Uint8Array(16),
    peerDeviceId: Uint8Array = new Uint8Array(16),
  ): void {
    const result = x3dhResponder(
      identity, signedPrekey, onetimePrekey, peerIdentity, peerEphemeral, ourDeviceId, peerDeviceId,
    );
    this._ratchet = DoubleRatchet.fromSharedSecret(
      result.sharedSecret, signedPrekey.privateKey, peerEphemeral, false,
    );
    this._establishedAt = Date.now();
    this._peerId = new Uint8Array(peerIdentity);
  }

  static fromSharedSecret(
    sharedSecret: Uint8Array,
    localDhPrivate: Uint8Array,
    remoteDhPublic: Uint8Array,
    isInitiator: boolean,
  ): Session {
    const s = new Session();
    s._ratchet = DoubleRatchet.fromSharedSecret(sharedSecret, localDhPrivate, remoteDhPublic, isInitiator);
    s._establishedAt = Date.now();
    return s;
  }

  encrypt(plaintext: Uint8Array, associatedData: Uint8Array = new Uint8Array(0)): Uint8Array {
    if (!this._ratchet) throw new Error('Session not established');
    return this._ratchet.ratchetEncrypt(plaintext, associatedData);
  }

  decrypt(ciphertext: Uint8Array, associatedData: Uint8Array = new Uint8Array(0)): Uint8Array {
    if (!this._ratchet) throw new Error('Session not established');
    return this._ratchet.ratchetDecrypt(ciphertext, associatedData);
  }
}
